### PR TITLE
(1) Add SHARP interface to M-CORE, (2) use send/recv to send train loss to the first rank instead of b-cast

### PR DESCRIPTION
**Pull request by erhoo82**
# What does this PR do ?

(1) Add SHARP interface to M-CORE in the communicator initialization.
(2) use send/recv to send train loss to the first rank instead of b-cast. This mitigates the communication overhead.

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks!

**Review by athitten — DISMISSED** [review]
LGTM, thank you!

**Comment by athitten** [comment]
@erhoo82 just remembered we want to get rid of the `torch broadcast` in `on_validation_epoch_end` as well right ? here: [torch.distributed.broadcast](https://github.com/NVIDIA/NeMo/pull/7793/files#diff-e93ccae74f4b67d341676afc9f3c7e2c50f751ec64df84eb3b2a86b62029ef76R985)

**Comment by erhoo82** [comment]
Somehow this was closed. Re-opend.

**Comment by erhoo82** [comment]
@athitten 
We need to keep the `torch.distirbuted.broadcast` in eval because this is the training termination condition for MLPerf.

**Comment by erhoo82** [comment]
Added some changes.

**Comment by erhoo82** [comment]
jenkins

**Comment by erhoo82** [comment]
@ericharper @athitten 
I think there is no issue with this PR right? Can we merge?

File: nemo/collections/nlp/parts/megatron_trainer_builder.py [hunk 1]
```diff
@@ -53,6 +53,7 @@ def _training_strategy(self) -> NLPDDPStrategy:
             no_ddp_communication_hook=True,
             gradient_as_bucket_view=self.cfg.model.gradient_as_bucket_view,
             find_unused_parameters=False,
+            sharp=cfg.model.get('sharp', False),
```
Comment (athitten) [inline nemo/collections/nlp/parts/megatron_trainer_builder.py:56]: @erhoo82 probably a typo. It should be `self.cfg.model.get` right ?

Comment (wdykas) [reply nemo/collections/nlp/parts/megatron_trainer_builder.py:56]: This breaks the running of this PR

**Comment by erhoo82** [comment]
Thanks!
Fixed the typo and resolved the conflict.

**Comment by ericharper** [comment]
jenkins

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Comment by ericharper** [comment]
jenkins

**Comment by erhoo82** [comment]
jenkins

**Review by layalir — APPROVED** [review]

**Pull request merged**